### ISSUE TITLE
CUDA_LAUNCH_BLOCKING=1 stucks sometime

### ISSUE THREAD

**Issue by acgtyrant** [pytorch/pytorch#9163]
## Issue description

This is the context of the `segment.py`:

    #!/usr/bin/env python

    import torch
    from torch import nn
    import torch.utils.data

    class DRN(nn.Module):
        def __init__(self):
            super(DRN, self).__init__()
            self.a = nn.Conv2d(3, 16, kernel_size=7)

        def forward(self, x):
            print('before DRN forward')
            return x

    if __name__ == '__main__':
        model = torch.nn.DataParallel(DRN()).cuda().train()
        input_ = torch.rand(2).cuda()
        print('before input')
        model(input_)
        print('end')

If I run `CUDA_VISIBLE_DEVICES=0,1 ./segment.py`, it will outputs

before input
before DRN forward
before DRN forward
end

However, if I run `CUDA_LAUNCH_BLOCKING=1 CUDA_VISIBLE_DEVICES=0,1  ./segment.py`, it will print `before input` only and stucks like below:

![2018-07-04-154901_1916x1058_scrot](https://user-images.githubusercontent.com/3921062/42263854-c97953ae-7fa1-11e8-9b9f-e5976388252f.png)

It very strange that if I change `rand(2)` to `rand(1)` or change `kernel__size=7` to `kernel_size=2`, it does not stuck again. So I describe this bug occurs "sometime".

I reproduce this bug in an machine which has two GTX 1080 too.

## System Info

Collecting environment information...
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Ubuntu 16.04.3 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.5) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.5
Is CUDA available: Yes
CUDA runtime version: 8.0.61
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 384.98
cuDNN version: Probably one of the following:
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.5.1.10
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.6.0.21
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn_static.a

Versions of relevant libraries:
[pip3] numpy (1.14.2)
[pip3] torch (0.4.0)
[pip3] torchvision (0.2.0)
[conda] Could not collect

**Comment by apaszke** [comment]
Yes this is expexted because of NCCL. You need to remove DataParallel or limit the process to a single GPU for debugging

**Comment by acgtyrant** [comment]
@apaszke Excuse me, how do you know this? I googled this problem many time, no one mention that.

**Comment by soumith** [comment]
@acgtyrant we asked some NVIDIA engineers. Apparently NCCL doesn't really like CUDA_LAUNCH_BLOCKING

**Comment by apaszke** [comment]
NCCL really needs to run multiple kernels on multiple GPUs simultaneously, or otherwise it will deadlock. This is why having 2 processes running DataParallels causes deadlocks (one starts a kernel on GPU0, another one on GPU1, you get a deadlock). Same issue happens here, because the CPU has to wait for GPU0 to return before it can queue the GPU1 kernel, but the GPU0 waits for GPU1.

**Comment by JiamingSuen** [comment]
What if a bug is only triggered only on multiple GPUs? In my case after training with multiple gpus for 3 hours the following error will occur:
```
cuda runtime error (59) : device-side assert triggered at /pytorch/torch/lib/THC/generic/THCStorage.c:184
Aborted (core dumped)
```

Any advice for debugging in this scenario?
Thanks.

**Comment by sjeaugey** [comment]
NCCL should work with CUDA_LAUNCH_BLOCKING, but only starting with CUDA 9 (the report mentions CUDA 8), since we can then launch kernels on multiple GPUs in a single cuda operation using cudaLaunchCooperativeKernelMultiDevice.

**Comment by dem123456789** [comment]
I think this should be mentioned somewhere because DataParallel stucks with zero warnings and error. It takes me one week to reach this page.
